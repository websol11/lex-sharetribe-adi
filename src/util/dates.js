import moment from 'moment-timezone/builds/moment-timezone-with-data-10-year-range.min';

/**
 * Input names for the DateRangePicker from react-dates.
 */
export const START_DATE = 'startDate';
export const END_DATE = 'endDate';

/**
 * Check if the browser's DateTimeFormat API supports time zones.
 *
 * @returns {Boolean} true if the browser returns current time zone.
 */
export const isTimeZoneSupported = () => {
  if (!Intl || typeof Intl === 'undefined' || typeof Intl.DateTimeFormat === 'undefined') {
    return false;
  }

  const dtf = new Intl.DateTimeFormat();
  if (typeof dtf === 'undefined' || typeof dtf.resolvedOptions === 'undefined') {
    return false;
  }
  return !!dtf.resolvedOptions().timeZone;
};

/**
 * Check if the given time zone key is valid.
 *
 * @param {String} timeZone time zone id, see:
 *   https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
 *
 * @returns {Boolean} true if the browser recognizes the key.
 */
export const isValidTimeZone = timeZone => {
  try {
    new Intl.DateTimeFormat('en-US', { timeZone }).format();
    return true;
  } catch (e) {
    return false;
  }
};

/**
 * Check that the given parameter is a Date object.
 *
 * @param {Date} object that should be a Date.
 *
 * @returns {boolean} true if given parameter is a Date object.
 */
export const isDate = d =>
  d && Object.prototype.toString.call(d) === '[object Date]' && !Number.isNaN(d.getTime());

/**
 * Check if the given parameters represent the same Date value (timestamps are compared)
 *
 * @param {Date} first param that should be a Date and it should have same timestamp as second param.
 * @param {Date} second param that should be a Date and it should have same timestamp as second param.
 *
 * @returns {boolean} true if given parameters have the same timestamp.
 */
export const isSameDate = (a, b) => a && isDate(a) && b && isDate(b) && a.getTime() === b.getTime();

/**
 * Compare is dateA is after dateB
 *
 * @param {Date} dateA date instance
 * @param {Date} dateB date instance
 *
 * @returns {Date} true if dateA is after dateB
 */
export const isAfterDate = (dateA, dateB) => {
  return moment(dateA).isAfter(moment(dateB));
};

////////////////////////////////////////////////////////////////////
// Manipulate time: time-of-day between different time zones etc. //
////////////////////////////////////////////////////////////////////

/**
 * Returns a new date, which indicates the same time of day in a given time zone
 * as given date is in local time zone
 *
 * @param {Date} date
 * @param {String} timeZone
 *
 * @returns {Date} date in given time zone
 */
export const timeOfDayFromLocalToTimeZone = (date, timeZone) => {
  return moment.tz(moment(date).format('YYYY-MM-DD HH:mm:ss'), timeZone).toDate();
};

/**
 * Returns a new date, which indicates the same time of day in a local time zone
 * as given date is in specified time zone
 *
 * @param {Date} date
 * @param {String} timeZone
 *
 * @returns {Date} date in given time zone
 */
export const timeOfDayFromTimeZoneToLocal = (date, timeZone) => {
  return moment(
    moment(date)
      .tz(timeZone)
      .format('YYYY-MM-DD HH:mm:ss')
  ).toDate();
};

/**
 * Get start of time unit (e.g. start of day)
 *
 * @param {Date} date date instance to be converted
 * @param {String} unit time-unit (e.g. "day")
 * @param {String} timeZone time zone id
 *
 * @returns {Date} date object converted to the start of given unit
 */
export const getStartOf = (date, unit, timeZone) => {
  const m = timeZone
    ? moment(date)
        .clone()
        .tz(timeZone)
    : moment(date).clone();

  return m.startOf(unit).toDate();
};

/**
 * Adds time-units to the date
 *
 * @param {Date} date date to be manipulated
 * @param {int} offset offset of time-units (e.g. "3" days)
 * @param {String} unit time-unit (e.g. "days")
 * @param {String} timeZone time zone name
 *
 * @returns {Date} date with given offset added
 */
export const addTime = (date, offset, unit, timeZone) => {
  const m = timeZone
    ? moment(date)
        .clone()
        .tz(timeZone)
    : moment(date).clone();
  return m.add(offset, unit).toDate();
};

/**
 * Subtract time-units from the date
 *
 * @param {Date} date date to be manipulated
 * @param {int} offset offset of time-units (e.g. "3" days)
 * @param {String} unit time-unit (e.g. "days")
 * @param {String} timeZone time zone name
 *
 * @returns {Date} date with given offset subtracted
 */
export const subtractTime = (date, offset, unit, timeZone) => {
  const m = timeZone
    ? moment(date)
        .clone()
        .tz(timeZone)
    : moment(date).clone();
  return m.subtract(offset, unit).toDate();
};

///////////////
// Durations //
///////////////

/**
 * Calculate the number of days between the given dates.
 * This uses moment#diff and, therefore, it just checks,
 * if there are 1000x60x60x24 milliseconds between date objects.
 *
 * Note: This should not be used for checking if the local date has
 *       changed between "2021-04-07 23:00" and "2021-04-08 05:00".
 *
 * @param {Date} startDate start of the time period
 * @param {Date} endDate end of the time period. NOTE: with daily
 * bookings, it is expected that this date is the exclusive end date,
 * i.e. the last day of the booking is the previous date of this end
 * date.
 *
 * @throws Will throw if the end date is before the start date
 * @returns {Number} number of days between the given dates
 */
export const daysBetween = (startDate, endDate) => {
  const days = moment(endDate).diff(startDate, 'days');
  if (days < 0) {
    throw new Error('End date cannot be before start date');
  }
  return days;
};

/**
 * Count the number of minutes between the given Date objects.
 *
 * @param {Date} startDate start of the time period
 * @param {Date} endDate end of the time period.
 *
 * @returns {Number} number of minutes between the given Date objects
 */
export const minutesBetween = (startDate, endDate) => {
  const minutes = moment(endDate).diff(startDate, 'minutes');
  return minutes;
};

/**
 * Calculate the difference between the given dates
 *
 * @param {Date} startDate start of the time period
 * @param {Date} endDate end of the time period.
 * @param {String} unit time unit. E.g. 'years'.
 * @param {String} useFloat Should return floating point numbers?
 *
 * @returns {Number} time difference between the given Date objects using given unit
 */
export const diffInTime = (startDate, endDate, unit, useFloat = false) => {
  return moment(startDate).diff(endDate, unit, useFloat);
};

////////////////////////////
// Parsing and formatting //
////////////////////////////

const getTimeZoneMaybe = timeZone => {
  if (timeZone) {
    if (!isTimeZoneSupported()) {
      throw new Error(`Your browser doesn't support time zones.`);
    }

    if (!isValidTimeZone(timeZone)) {
      throw new Error(`Given time zone key (${timeZone}) is not valid.`);
    }
    return { timeZone };
  }
  return {};
};

/**
 * Format the given date. Printed string depends on how close the date is the current day.
 * E.g. "Today, 9:10 PM", "Sun 6:02 PM", "Jul 20, 6:02 PM", "Jul 20 2020, 6:02 PM"
 *
 * @param {Date} date Date to be formatted
 * @param {Object} intl Intl object from react-intl
 * @param {String} todayString translation for the current day
 * @param {Object} [opts] options. Can be used to pass in timeZone. It should represent IANA time zone key.
 *
 * @returns {String} formatted date
 */
export const formatDateWithProximity = (date, intl, todayString, opts = {}) => {
  const paramsValid = intl && date instanceof Date && typeof todayString === 'string';
  if (!paramsValid) {
    throw new Error(`Invalid params for formatDate: (${date}, ${intl}, ${todayString})`);
  }

  // If timeZone parameter is set, use it as formatting option
  const { timeZone } = opts;
  const timeZoneMaybe = getTimeZoneMaybe(timeZone);

  // By default we can use moment() directly but in tests we need to use a specific dates.
  // Tests inject now() function to intl wich returns predefined date
  const now = intl.now ? moment(intl.now()) : moment();

  // isSame: if the two moments have different time zones, the time zone of the first moment will be used for the comparison.
  const localizedNow = timeZoneMaybe.timeZone ? now.tz(timeZone) : now;

  if (localizedNow.isSame(date, 'day')) {
    // e.g. "Today, 9:10 PM"
    const formattedTime = intl.formatDate(date, {
      hour: 'numeric',
      minute: 'numeric',
      ...timeZoneMaybe,
    });
    return `${todayString}, ${formattedTime}`;
  } else if (localizedNow.isSame(date, 'week')) {
    // e.g.
    // en-US: "Sun 6:02 PM"
    // en-GB: "Sun 18:02"
    // fr-FR: "dim. 18:02"
    return intl.formatDate(date, {
      weekday: 'short',
      hour: 'numeric',
      minute: 'numeric',
      ...timeZoneMaybe,
    });
  } else if (localizedNow.isSame(date, 'year')) {
    // e.g.
    // en-US: "Jul 20, 6:02 PM"
    // en-GB: "20 Jul, 18:02"
    // fr-FR: "20 juil., 18:02"
    return intl.formatDate(date, {
      month: 'short',
      day: 'numeric',
      hour: 'numeric',
      minute: 'numeric',
      ...timeZoneMaybe,
    });
  } else {
    // e.g.
    // en-US: "Jul 20, 2020, 6:02 PM"
    // en-GB: "20 Jul 2020, 18:02"
    // fr-FR: "20 juil. 2020, 18:02"
    return intl.formatDate(date, {
      year: 'numeric',
      month: 'short',
      day: 'numeric',
      hour: 'numeric',
      minute: 'numeric',
      ...timeZoneMaybe,
    });
  }
};

/**
 * Formats date to into multiple different ways:
 * - date "Mar 24"
 * - time "8:07 PM"
 * - dateAndTime: "Mar 24, 8:07 PM"
 *
 * If date is on different year, it will show it.
 *
 * @param {Date} date to be formatted
 * @param {Object} intl Intl object from react-intl
 * @param {Object} [opts] options. Can be used to pass in timeZone. It should represent IANA time zone key.
 *
 * @returns {Object} "{ date, time, dateAndTime }"
 */
export const formatDateIntoPartials = (date, intl, opts = {}) => {
  // If timeZone parameter is set, use it as formatting option
  const { timeZone } = opts;
  const timeZoneMaybe = getTimeZoneMaybe(timeZone);

  // By default we can use moment() directly but in tests we need to use a specific dates.
  // Tests inject now() function to intl wich returns predefined date
  const now = intl.now ? moment(intl.now()) : moment();

  // isSame: if the two moments have different time zones, the time zone of the first moment will be used for the comparison.
  const localizedNow = timeZoneMaybe.timeZone ? now.tz(timeZone) : now;
  const yearMaybe = localizedNow.isSame(date, 'year') ? {} : { year: 'numeric' };

  return {
    date: intl.formatDate(date, {
      month: 'short',
      day: 'numeric',
      ...yearMaybe,
      ...timeZoneMaybe,
    }),
    time: intl.formatDate(date, {
      hour: 'numeric',
      minute: 'numeric',
      ...timeZoneMaybe,
    }),
    dateAndTime: intl.formatDate(date, {
      ...yearMaybe,
      month: 'short',
      day: 'numeric',
      hour: 'numeric',
      minute: 'numeric',
      ...timeZoneMaybe,
    }),
  };
};

/**
 * Parses given date string in ISO8601 format('YYYY-MM-DD') to date in
 * the given time zone.
 *
 * This is used in search when filtering by time-based availability.
 *
 * Example:
 * ('2020-04-15', 'Etc/UTC') => new Date('2020-04-15T00:00:00.000Z')
 * ('2020-04-15', 'Europe/Helsinki') => new Date('2020-04-14T21:00:00.000Z')
 *
 * @param {String} dateString in 'YYYY-MM-DD' format
 * @param {String} [timeZone] time zone id, see:
 *   https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
 *
 * @returns {Date} date
 */
export const parseDateFromISO8601 = (dateString, timeZone = null) => {
  return timeZone
    ? moment.tz(dateString, timeZone).toDate()
    : moment(dateString, 'YYYY-MM-DD').toDate();
};

/**
 * Converts date to string ISO8601 format ('YYYY-MM-DD').
 * This string is used e.g. in urlParam.
 *
 * @param {Date} date
 * @param {String} [timeZone] time zone id, see:
 *   https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
 *
 * @returns {String} string in 'YYYY-MM-DD' format
 */
export const stringifyDateToISO8601 = (date, timeZone = null) => {
  return timeZone
    ? moment(date)
        .tz(timeZone)
        .format('YYYY-MM-DD')
    : moment(date).format('YYYY-MM-DD');
};

//////////
// Misc //
//////////

/**
 * Format the given date to month id/string: 'YYYY-MM'.
 *
 * @param {Date} date to be formatted
 * @param {String} [timeZone] time zone name (optional parameter).
 *
 * @returns {String} formatted month string
 */
export const monthIdString = (date, timeZone = null) => {
  return timeZone
    ? moment(date)
        .tz(timeZone)
        .format('YYYY-MM')
    : moment(date).format('YYYY-MM');
};

/**
 * Formats string ('YYYY-MM-DD') in given time zone to format ('0000-00-00T00:00:00.000Z') and adds one day.
 * This is used as end date of the search query.
 * One day must be added because end of the availability is exclusive in API.
 *
 * @param {String} string in 'YYYY-MM-DD'format
 * @param {String} timeZone time zone name.
 *
 * @returns {String} string in '0000-00-00T00:00:00.000Z' format
 */
export const getExclusiveEndDate = (dateString, timeZone) => {
  return moment
    .tz(dateString, timeZone)
    .add(1, 'days')
    .startOf('day')
    .toDate();
};

/**
 * Convert local date for API.
 * Date given by browser
 * ("Fri Mar 30 2018 12:00:00 GMT-1100 (SST)" aka "Fri Mar 30 2018 23:00:00 GMT+0000 (UTC)")
 * must be modified so that API will get correct moment also in UTC.
 * We achieve this by adding timezone offset to local date / timestamp.
 *
 * The resulting timestamp for the API is:
 * localTimestamp.add(timezoneoffset)
 * In eg. Fri Mar 30 2018 23:00:00 GMT-1100 (SST) aka "Fri Mar 30 2018 12:00:00 GMT+0000 (UTC)"
 *
 * @param {Date} date is a local date object
 *
 * @returns {Date} date (given by API as UTC 00:00) converted back to local noon.
 */
export const dateFromLocalToAPI = date => {
  const timezoneDiffInMinutes = moment(date).utcOffset();
  const momentInLocalTimezone = moment(date).add(timezoneDiffInMinutes, 'minutes');

  return momentInLocalTimezone.toDate();
};
