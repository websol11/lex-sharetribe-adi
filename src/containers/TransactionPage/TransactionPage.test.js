import React from 'react';
import {
  createCurrentUser,
  createListing,
  createTransaction,
  createUser,
  fakeIntl,
} from '../../util/test-data';
import { renderShallow } from '../../util/test-helpers';
import { TRANSITION_CONFIRM_PAYMENT } from '../../util/transaction';
import { TransactionPageComponent } from './TransactionPage';

const noop = () => null;

describe('TransactionPage - Sale', () => {
  it('matches snapshot', () => {
    const txId = 'tx-sale-1';
    const start = new Date(Date.UTC(2017, 5, 10));
    const end = new Date(Date.UTC(2017, 5, 13));
    const transaction = createTransaction({
      id: txId,
      lastTransition: TRANSITION_CONFIRM_PAYMENT,
      listing: createListing('listing1'),
      customer: createUser('customer1'),
      provider: createUser('provider1'),
    });

    const props = {
      params: { id: txId },
      transactionRole: 'provider',

      currentUser: createCurrentUser('provider1'),
      markReceivedFromPurchasedInProgress: false,
      onMarkReceivedFromPurchased: noop,
      markDeliveredInProgress: false,
      onMarkDelivered: noop,
      markReceivedInProgress: false,
      onMarkReceived: noop,
      disputeInProgress: false,
      onDispute: noop,
      scrollingDisabled: false,
      callSetInitialValues: noop,
      transaction,
      totalMessages: 0,
      totalMessagePages: 0,
      oldestMessagePageFetched: 0,
      messages: [],
      sendMessageInProgress: false,
      onInitializeCardPaymentData: noop,
      onShowMoreMessages: noop,
      onSendMessage: noop,
      onResetForm: noop,
      onFetchTransactionLineItems: noop,
      fetchLineItemsInProgress: false,
      intl: fakeIntl,

      location: {
        pathname: `/sale/${txId}/details`,
        search: '',
        hash: '',
      },
      history: {
        push: () => console.log('HistoryPush called'),
      },
    };

    const tree = renderShallow(<TransactionPageComponent {...props} />);
    expect(tree).toMatchSnapshot();
  });
});

describe('TransactionPage - Order', () => {
  it('matches snapshot', () => {
    const txId = 'tx-order-1';
    const start = new Date(Date.UTC(2017, 5, 10));
    const end = new Date(Date.UTC(2017, 5, 13));

    const transaction = createTransaction({
      id: txId,
      lastTransition: TRANSITION_CONFIRM_PAYMENT,
      listing: createListing('listing1'),
      customer: createUser('customer1'),
      provider: createUser('provider1'),
    });

    const props = {
      params: { id: txId },
      transactionRole: 'customer',

      currentUser: createCurrentUser('customer1'),
      totalMessages: 0,
      totalMessagePages: 0,
      oldestMessagePageFetched: 0,
      messages: [],
      fetchMessagesInProgress: false,
      sendMessageInProgress: false,
      scrollingDisabled: false,
      callSetInitialValues: noop,
      transaction,
      onInitializeCardPaymentData: noop,
      onShowMoreMessages: noop,
      onSendMessage: noop,
      onResetForm: noop,
      onFetchTransactionLineItems: noop,
      fetchLineItemsInProgress: false,
      intl: fakeIntl,

      markReceivedFromPurchasedInProgress: false,
      onMarkReceivedFromPurchased: noop,
      markDeliveredInProgress: false,
      onMarkDelivered: noop,
      markReceivedInProgress: false,
      onMarkReceived: noop,
      disputeInProgress: false,
      onDispute: noop,

      location: {
        pathname: `/order/${txId}/details`,
        search: '',
        hash: '',
      },
      history: {
        push: () => console.log('HistoryPush called'),
      },
    };

    const tree = renderShallow(<TransactionPageComponent {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
