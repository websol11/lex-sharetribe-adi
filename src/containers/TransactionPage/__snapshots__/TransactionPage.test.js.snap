// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionPage - Order matches snapshot 1`] = `
<Page
  scrollingDisabled={false}
  title="TransactionPage.title"
>
  <LayoutSingleColumn
    className={null}
    rootClassName={null}
  >
    <LayoutWrapperTopbar
      className={null}
      rootClassName={null}
    >
      <withRouter(Connect(TopbarContainerComponent)) />
    </LayoutWrapperTopbar>
    <LayoutWrapperMain
      className={null}
      rootClassName={null}
    >
      <div
        className="root"
      >
        <injectIntl(TransactionPanelComponent)
          className="tabContent tabContentVisible"
          currentUser={
            Object {
              "attributes": Object {
                "banned": false,
                "deleted": false,
                "email": "customer1@example.com",
                "emailVerified": true,
                "profile": Object {
                  "abbreviatedName": "customer1 abbreviated name",
                  "displayName": "customer1 display name",
                  "firstName": "customer1 first name",
                  "lastName": "customer1 last name",
                },
              },
              "id": UUID {
                "_sdkType": "UUID",
                "uuid": "customer1",
              },
              "type": "currentUser",
            }
          }
          fetchLineItemsError={null}
          fetchLineItemsInProgress={false}
          fetchMessagesError={null}
          fetchMessagesInProgress={false}
          fetchTimeSlotsError={null}
          initialMessageFailed={false}
          leaveReviewProps={
            Object {
              "buttonText": "TransactionPage.leaveReview.actionButton",
              "error": undefined,
              "errorText": "TransactionPage.leaveReview.actionError",
              "inProgress": undefined,
              "onTransition": [Function],
            }
          }
          lineItems={null}
          markDeliveredProps={
            Object {
              "buttonText": "TransactionPage.markDelivered.actionButton",
              "error": null,
              "errorText": "TransactionPage.markDelivered.actionError",
              "inProgress": false,
              "onTransition": [Function],
            }
          }
          markReceivedFromPurchasedProps={
            Object {
              "buttonText": "TransactionPage.markReceivedFromPurchased.actionButton",
              "error": null,
              "errorText": "TransactionPage.markReceivedFromPurchased.actionError",
              "inProgress": false,
              "onTransition": [Function],
            }
          }
          markReceivedProps={
            Object {
              "buttonText": "TransactionPage.markReceived.actionButton",
              "error": null,
              "errorText": "TransactionPage.markReceived.actionError",
              "inProgress": false,
              "onTransition": [Function],
            }
          }
          messages={Array []}
          oldestMessagePageFetched={0}
          onFetchTransactionLineItems={[Function]}
          onOpenDisputeModal={[Function]}
          onOpenReviewModal={[Function]}
          onSendMessage={[Function]}
          onShowMoreMessages={[Function]}
          onSubmitOrderRequest={[Function]}
          savePaymentMethodFailed={false}
          sendMessageError={null}
          sendMessageInProgress={false}
          timeSlots={null}
          totalMessagePages={0}
          transaction={
            Object {
              "attributes": Object {
                "createdAt": 2017-05-01T00:00:00.000Z,
                "lastTransition": "transition/confirm-payment",
                "lastTransitionedAt": 2017-06-01T00:00:00.000Z,
                "lineItems": Array [
                  Object {
                    "code": "line-item/day",
                    "includeFor": Array [
                      "customer",
                      "provider",
                    ],
                    "lineTotal": Money {
                      "_sdkType": "Money",
                      "amount": 1000,
                      "currency": "USD",
                    },
                    "quantity": "1",
                    "reversal": false,
                    "unitPrice": Money {
                      "_sdkType": "Money",
                      "amount": 1000,
                      "currency": "USD",
                    },
                  },
                  Object {
                    "code": "line-item/provider-commission",
                    "includeFor": Array [
                      "provider",
                    ],
                    "lineTotal": Money {
                      "_sdkType": "Money",
                      "amount": -100,
                      "currency": "USD",
                    },
                    "reversal": false,
                    "unitPrice": Money {
                      "_sdkType": "Money",
                      "amount": -100,
                      "currency": "USD",
                    },
                  },
                ],
                "payinTotal": Money {
                  "_sdkType": "Money",
                  "amount": 1000,
                  "currency": "USD",
                },
                "payoutTotal": Money {
                  "_sdkType": "Money",
                  "amount": 900,
                  "currency": "USD",
                },
                "transitions": Array [
                  Object {
                    "by": "customer",
                    "createdAt": 2017-05-01T00:00:00.000Z,
                    "transition": "transition/request-payment",
                  },
                  Object {
                    "by": "customer",
                    "createdAt": 2017-05-01T00:00:01.000Z,
                    "transition": "transition/confirm-payment",
                  },
                ],
              },
              "booking": null,
              "customer": Object {
                "attributes": Object {
                  "banned": false,
                  "deleted": false,
                  "profile": Object {
                    "abbreviatedName": "TT",
                    "displayName": "customer1 display name",
                  },
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "customer1",
                },
                "type": "user",
              },
              "id": UUID {
                "_sdkType": "UUID",
                "uuid": "tx-order-1",
              },
              "listing": Object {
                "attributes": Object {
                  "deleted": false,
                  "description": "listing1 description",
                  "geolocation": LatLng {
                    "_sdkType": "LatLng",
                    "lat": 40,
                    "lng": 60,
                  },
                  "price": Money {
                    "_sdkType": "Money",
                    "amount": 5500,
                    "currency": "USD",
                  },
                  "publicData": Object {},
                  "state": "published",
                  "title": "listing1 title",
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "listing1",
                },
                "type": "listing",
              },
              "provider": Object {
                "attributes": Object {
                  "banned": false,
                  "deleted": false,
                  "profile": Object {
                    "abbreviatedName": "TT",
                    "displayName": "provider1 display name",
                  },
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "provider1",
                },
                "type": "user",
              },
              "reviews": Array [],
              "type": "transaction",
            }
          }
          transactionRole="customer"
        />
      </div>
      <injectIntl(ReviewModal)
        id="ReviewOrderModal"
        isOpen={false}
        onCloseModal={[Function]}
        onSubmitReview={[Function]}
        reviewSent={false}
        revieweeName={
          <UserDisplayName
            bannedUserDisplayName={null}
            className={null}
            deletedUserDisplayName={null}
            intl={
              Object {
                "formatDate": [Function],
                "formatMessage": [Function],
                "formatNumber": [Function],
                "formatPlural": [Function],
                "formatRelativeTime": [Function],
                "formatTime": [Function],
                "now": [Function],
              }
            }
            rootClassName={null}
            user={
              Object {
                "attributes": Object {
                  "banned": false,
                  "deleted": false,
                  "profile": Object {
                    "abbreviatedName": "TT",
                    "displayName": "provider1 display name",
                  },
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "provider1",
                },
                "type": "user",
              }
            }
          />
        }
      />
      <injectIntl(DisputeModal)
        disputeError={null}
        disputeInProgress={false}
        disputeSubmitted={false}
        id="DisputeOrderModal"
        isOpen={false}
        onCloseModal={[Function]}
        onDisputeOrder={[Function]}
      />
    </LayoutWrapperMain>
    <LayoutWrapperFooter
      className="footer"
      rootClassName={null}
    >
      <injectIntl(Footer) />
    </LayoutWrapperFooter>
  </LayoutSingleColumn>
</Page>
`;

exports[`TransactionPage - Sale matches snapshot 1`] = `
<Page
  scrollingDisabled={false}
  title="TransactionPage.title"
>
  <LayoutSingleColumn
    className={null}
    rootClassName={null}
  >
    <LayoutWrapperTopbar
      className={null}
      rootClassName={null}
    >
      <withRouter(Connect(TopbarContainerComponent)) />
    </LayoutWrapperTopbar>
    <LayoutWrapperMain
      className={null}
      rootClassName={null}
    >
      <div
        className="root"
      >
        <injectIntl(TransactionPanelComponent)
          className="tabContent tabContentVisible"
          currentUser={
            Object {
              "attributes": Object {
                "banned": false,
                "deleted": false,
                "email": "provider1@example.com",
                "emailVerified": true,
                "profile": Object {
                  "abbreviatedName": "provider1 abbreviated name",
                  "displayName": "provider1 display name",
                  "firstName": "provider1 first name",
                  "lastName": "provider1 last name",
                },
              },
              "id": UUID {
                "_sdkType": "UUID",
                "uuid": "provider1",
              },
              "type": "currentUser",
            }
          }
          fetchLineItemsError={null}
          fetchLineItemsInProgress={false}
          fetchMessagesError={null}
          fetchTimeSlotsError={null}
          initialMessageFailed={false}
          leaveReviewProps={
            Object {
              "buttonText": "TransactionPage.leaveReview.actionButton",
              "error": undefined,
              "errorText": "TransactionPage.leaveReview.actionError",
              "inProgress": undefined,
              "onTransition": [Function],
            }
          }
          lineItems={null}
          markDeliveredProps={
            Object {
              "buttonText": "TransactionPage.markDelivered.actionButton",
              "error": null,
              "errorText": "TransactionPage.markDelivered.actionError",
              "inProgress": false,
              "onTransition": [Function],
            }
          }
          markReceivedFromPurchasedProps={
            Object {
              "buttonText": "TransactionPage.markReceivedFromPurchased.actionButton",
              "error": null,
              "errorText": "TransactionPage.markReceivedFromPurchased.actionError",
              "inProgress": false,
              "onTransition": [Function],
            }
          }
          markReceivedProps={
            Object {
              "buttonText": "TransactionPage.markReceived.actionButton",
              "error": null,
              "errorText": "TransactionPage.markReceived.actionError",
              "inProgress": false,
              "onTransition": [Function],
            }
          }
          messages={Array []}
          oldestMessagePageFetched={0}
          onFetchTransactionLineItems={[Function]}
          onOpenDisputeModal={[Function]}
          onOpenReviewModal={[Function]}
          onSendMessage={[Function]}
          onShowMoreMessages={[Function]}
          onSubmitOrderRequest={[Function]}
          savePaymentMethodFailed={false}
          sendMessageError={null}
          sendMessageInProgress={false}
          timeSlots={null}
          totalMessagePages={0}
          transaction={
            Object {
              "attributes": Object {
                "createdAt": 2017-05-01T00:00:00.000Z,
                "lastTransition": "transition/confirm-payment",
                "lastTransitionedAt": 2017-06-01T00:00:00.000Z,
                "lineItems": Array [
                  Object {
                    "code": "line-item/day",
                    "includeFor": Array [
                      "customer",
                      "provider",
                    ],
                    "lineTotal": Money {
                      "_sdkType": "Money",
                      "amount": 1000,
                      "currency": "USD",
                    },
                    "quantity": "1",
                    "reversal": false,
                    "unitPrice": Money {
                      "_sdkType": "Money",
                      "amount": 1000,
                      "currency": "USD",
                    },
                  },
                  Object {
                    "code": "line-item/provider-commission",
                    "includeFor": Array [
                      "provider",
                    ],
                    "lineTotal": Money {
                      "_sdkType": "Money",
                      "amount": -100,
                      "currency": "USD",
                    },
                    "reversal": false,
                    "unitPrice": Money {
                      "_sdkType": "Money",
                      "amount": -100,
                      "currency": "USD",
                    },
                  },
                ],
                "payinTotal": Money {
                  "_sdkType": "Money",
                  "amount": 1000,
                  "currency": "USD",
                },
                "payoutTotal": Money {
                  "_sdkType": "Money",
                  "amount": 900,
                  "currency": "USD",
                },
                "transitions": Array [
                  Object {
                    "by": "customer",
                    "createdAt": 2017-05-01T00:00:00.000Z,
                    "transition": "transition/request-payment",
                  },
                  Object {
                    "by": "customer",
                    "createdAt": 2017-05-01T00:00:01.000Z,
                    "transition": "transition/confirm-payment",
                  },
                ],
              },
              "booking": null,
              "customer": Object {
                "attributes": Object {
                  "banned": false,
                  "deleted": false,
                  "profile": Object {
                    "abbreviatedName": "TT",
                    "displayName": "customer1 display name",
                  },
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "customer1",
                },
                "type": "user",
              },
              "id": UUID {
                "_sdkType": "UUID",
                "uuid": "tx-sale-1",
              },
              "listing": Object {
                "attributes": Object {
                  "deleted": false,
                  "description": "listing1 description",
                  "geolocation": LatLng {
                    "_sdkType": "LatLng",
                    "lat": 40,
                    "lng": 60,
                  },
                  "price": Money {
                    "_sdkType": "Money",
                    "amount": 5500,
                    "currency": "USD",
                  },
                  "publicData": Object {},
                  "state": "published",
                  "title": "listing1 title",
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "listing1",
                },
                "type": "listing",
              },
              "provider": Object {
                "attributes": Object {
                  "banned": false,
                  "deleted": false,
                  "profile": Object {
                    "abbreviatedName": "TT",
                    "displayName": "provider1 display name",
                  },
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "provider1",
                },
                "type": "user",
              },
              "reviews": Array [],
              "type": "transaction",
            }
          }
          transactionRole="provider"
        />
      </div>
      <injectIntl(ReviewModal)
        id="ReviewOrderModal"
        isOpen={false}
        onCloseModal={[Function]}
        onSubmitReview={[Function]}
        reviewSent={false}
        revieweeName={
          <UserDisplayName
            bannedUserDisplayName={null}
            className={null}
            deletedUserDisplayName={null}
            intl={
              Object {
                "formatDate": [Function],
                "formatMessage": [Function],
                "formatNumber": [Function],
                "formatPlural": [Function],
                "formatRelativeTime": [Function],
                "formatTime": [Function],
                "now": [Function],
              }
            }
            rootClassName={null}
            user={
              Object {
                "attributes": Object {
                  "banned": false,
                  "deleted": false,
                  "profile": Object {
                    "abbreviatedName": "TT",
                    "displayName": "customer1 display name",
                  },
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "customer1",
                },
                "type": "user",
              }
            }
          />
        }
      />
      <injectIntl(DisputeModal)
        disputeError={null}
        disputeInProgress={false}
        disputeSubmitted={false}
        id="DisputeOrderModal"
        isOpen={false}
        onCloseModal={[Function]}
        onDisputeOrder={[Function]}
      />
    </LayoutWrapperMain>
    <LayoutWrapperFooter
      className="footer"
      rootClassName={null}
    >
      <injectIntl(Footer) />
    </LayoutWrapperFooter>
  </LayoutSingleColumn>
</Page>
`;
